changelog/3149.bugfix

```diff
@@ -0,0 +1 @@
+Fix doctest failure output so that it will continue to run despite the first failure.
```

Comment:
Please rename this file to `3149.feature` and l suggest to update the text to:

```
Doctest runs now show multiple failures for each doctest snippet, instead of stopping at the first failure.
```

Reply:
I'll change it to a feature.  I may need help understanding the failed tests though.  It seems like with --pdb I'm experiencing a timeout issue with the py36-pexpect test.  Can you tell me how the pdb stdout/stdin capture would work in the case where there's an assert statement in the doctest (in pty_spawn.py's test_pdb_interaction_doctest)?

I think in such case should I make it fail like before (raise immediately inside the DoctestItem's runtest)?  How do I test if pdb is turned on?

Reply:
This test ensures that passing `--pdb` to pytest will make `pdb` stop at the point of failure inside the doctests when an exception happens.

It is easier to understand this if you manually replicate what the test is doing. 

Create a `test_docs.py` file:

```python
def function_1():
    '''
    >>> i = 0
    >>> assert i == 1
    '''
```

If you execute this on `master` using this command-line:

```
pytest --pdb --doctest-modules test_docs.py
```

You will notice pytest will enter `pdb` at the point the exception happened:

```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
003
004     >>> i = 0
005     >>> assert i == 1
UNEXPECTED EXCEPTION: AssertionError()
Traceback (most recent call last):

  File "C:\Users\Bruno\AppData\Local\Programs\Python\Python36\lib\doctest.py", line 1330, in __run
    compileflags, 1), test.globs)

  File "<doctest test_docs.function_1[1]>", line 1, in <module>

AssertionError

C:\pytest\.tmp\test_docs.py:5: UnexpectedException
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> <doctest test_docs.function_1[1]>(1)<module>()
(Pdb) 
```

If you enter `i` (to inspect the value of `i`), you get the expected value:

```
(Pdb) i
0
```

Using your branch however, we get into a frame which is not the point where the original error occurred, but in the point where you raise the `DoctestFailure(failures)` exception inside `runtest()`:

```
(Pdb) w
  c:\pytest\_pytest\runner.py(189)__init__()
-> self.result = func()
  c:\pytest\_pytest\runner.py(175)<lambda>()
-> return CallInfo(lambda: ihook(item=item, **kwds), when=when)
  c:\pytest\.env36\lib\site-packages\pluggy\hooks.py(256)__call__()
-> return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  c:\pytest\.env36\lib\site-packages\pluggy\manager.py(43)_hookexec()
-> return self._inner_hookexec(hook, methods, kwargs)
  c:\pytest\.env36\lib\site-packages\pluggy\manager.py(37)<lambda>()
-> firstresult=hook.spec_opts.get('firstresult'),
  c:\pytest\.env36\lib\site-packages\pluggy\callers.py(201)_multicall()
-> return outcome.get_result()
  c:\pytest\.env36\lib\site-packages\pluggy\callers.py(76)get_result()
-> raise ex[1].with_traceback(ex[2])
  c:\pytest\.env36\lib\site-packages\pluggy\callers.py(180)_multicall()
-> res = hook_impl.function(*args)
  c:\pytest\_pytest\runner.py(106)pytest_runtest_call()
-> item.runtest()
> c:\pytest\_pytest\doctest.py(170)runtest()
-> raise DoctestFailure(failures)
(Pdb)
```

So sending `i` will not output the expected `0` in the console:

```
(Pdb) i
*** NameError: name 'i' is not defined
```

Which is why the test is failing on Travis: it is timing out because it is waiting for the process to output `0` , which never happens.

---

I've tried to fix this test with this patch:

```diff
diff --git a/_pytest/debugging.py b/_pytest/debugging.py
index d7dca78..d961327 100644
--- a/_pytest/debugging.py
+++ b/_pytest/debugging.py
@@ -97,11 +97,15 @@ def _enter_pdb(node, excinfo, rep):
 def _postmortem_traceback(excinfo):
     # A doctest.UnexpectedException is not useful for post_mortem.
     # Use the underlying exception instead:
+    from _pytest.doctest import DoctestFailure
     from doctest import UnexpectedException
-    if isinstance(excinfo.value, UnexpectedException):
-        return excinfo.value.exc_info[2]
-    else:
-        return excinfo._excinfo[2]
+    if isinstance(excinfo.value, DoctestFailure):
+        failures = excinfo.value.failures
+        for failure in failures:
+            if isinstance(failure, UnexpectedException):
+                return failure.exc_info[2]
+
+    return excinfo._excinfo[2]


 def _find_last_non_hidden_frame(stack):
diff --git a/_pytest/doctest.py b/_pytest/doctest.py
index 6fe644e..d7cd74a 100644
--- a/_pytest/doctest.py
+++ b/_pytest/doctest.py
@@ -124,10 +124,18 @@ def _init_runner_class():
             pass

         def report_failure(self, out, test, example, got):
-            out.append(DoctestFailureContainer(test, example, got))
+            from doctest import DocTestFailure
+            try:
+                raise DocTestFailure(test, example, got)
+            except DoctestFailure as e:
+                out.append(e)

         def report_unexpected_exception(self, out, test, example, exc_info):
-            out.append(DoctestUnexpectedExceptionContainer(test, example, exc_info))
+            from doctest import UnexpectedException
+            try:
+                raise UnexpectedException(test, example, exc_info)
+            except UnexpectedException as e:
+                out.append(e)

     return PytestDoctestRunner

@@ -182,7 +190,7 @@ class DoctestItem(pytest.Item):
                     lineno = None
                 else:
                     lineno = test.lineno + example.lineno + 1
-                message = failure.NAME
+                message = type(failure).__name__
                 reprlocation = ReprFileLocation(filename, lineno, message)
                 checker = _get_checker()
                 report_choice = _get_report_choice(self.config.getoption("doctestreport"))
```

Unfortunately this did not solve the problem. 😕 

I don't have time to investigate this right now and plan to take another look early next week, but I thought I would post this if you want to investigate further yourself. 👍

Reply:
I made some changes in my local checkout so that if pdb is set, I'll raise an exception inside the runtest call immediately instead of running through.  I think this makes more sense when getting into pdb mode.  When I run tox -e py36-pexpect/py27-pexpect seems to pass now, but I'm still a little confused about running tests outside of tox to test your example for the local assert.

Can you tell me what's the proper way to run pytest in the development environment BTW?  I tried to do a "python setup.py develop" and run pytest on the module that contains the doctest but it can't seem to find the test.

BTW, since my forked master is modified, do you want me to fork again from a clean view and apply my changes to the features branch?  Or is there a more correct way to do that?

Reply:
> Can you tell me what's the proper way to run pytest in the development environment BTW? I tried to do a "python setup.py develop" and run pytest on the module that contains the doctest but it can't seem to find the test.

You need to also install the dependencies, you can check what those are for that environment by looking at the `tox.ini` file:

https://github.com/pytest-dev/pytest/blob/570edb466b90795845b71795e87b4b6186cd85c6/tox.ini#L66-L71

So for that environment you probably only need to install `pexpect` (otherwise the test is skipped).

> BTW, since my forked master is modified, do you want me to fork again from a clean view and apply my changes to the features branch? Or is there a more correct way to do that?

If you have the time this would be the cleaner approach, thanks!